src/Model_Hydro/Hydro_BoundaryCondition_Diode.cpp

```diff
@@ -0,0 +1,618 @@
+   if ( NVar_Der != 0  &&  TDerVarList == NULL )
+      Aux_Error( ERROR_INFO, "NVar_Der = %d != 0, TDerVarList == NULL !!\n", NVar_Der );
+#  endif // #ifdef GAMER_DEBUG
+
+
+// set the boundary values at different boundary faces
+   switch ( BC_Face )
+   {
+      case 0:  BC_Diode_xm( Array, NVar_Flu, TFluVarIdxList, NVar_Der, TDerVarList, GhostSize,
+                            ArraySizeX, ArraySizeY, ArraySizeZ, Idx_Start, Idx_End );   break;
+
+      case 1:  BC_Diode_xp( Array, NVar_Flu, TFluVarIdxList, NVar_Der, TDerVarList, GhostSize,
+                            ArraySizeX, ArraySizeY, ArraySizeZ, Idx_Start, Idx_End );   break;
+
+      case 2:  BC_Diode_ym( Array, NVar_Flu, TFluVarIdxList, NVar_Der, TDerVarList, GhostSize,
+                            ArraySizeX, ArraySizeY, ArraySizeZ, Idx_Start, Idx_End );   break;
+
+      case 3:  BC_Diode_yp( Array, NVar_Flu, TFluVarIdxList, NVar_Der, TDerVarList, GhostSize,
+                            ArraySizeX, ArraySizeY, ArraySizeZ, Idx_Start, Idx_End );   break;
+
+      case 4:  BC_Diode_zm( Array, NVar_Flu, TFluVarIdxList, NVar_Der, TDerVarList, GhostSize,
+                            ArraySizeX, ArraySizeY, ArraySizeZ, Idx_Start, Idx_End );   break;
+
+      case 5:  BC_Diode_zp( Array, NVar_Flu, TFluVarIdxList, NVar_Der, TDerVarList, GhostSize,
+                            ArraySizeX, ArraySizeY, ArraySizeZ, Idx_Start, Idx_End );   break;
+
+      default: Aux_Error( ERROR_INFO, "incorrect boundary face (%d) !!\n", BC_Face );
+   }
+
+} // FUNCTION : Hydro_BoundaryCondition_Diode
+
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  BC_Diode_xm
+// Description :  Set the diode B.C. at the -x boundary
+//
+// Note        :  1. Work for Hydro_BoundaryCondition_Diode()
+//                2. Use the input parameter "NVar_Flu" and "TFluVarIdxList" to control the target fluid variables
+//
+// Parameter   :  See Hydro_BoundaryCondition_Diode()
+//
+// Return      :  Array
+//-------------------------------------------------------------------------------------------------------
+void BC_Diode_xm( real *Array, const int NVar_Flu, const int TFluVarIdxList[], const int NVar_Der,
+                  const long TDerVarList[], const int GhostSize, const int ArraySizeX, const int ArraySizeY,
+                  const int ArraySizeZ, const int Idx_Start[], const int Idx_End[] )
+{
+
+   const int i_ref = GhostSize;    // reference i index
+   int TFluVarIdx, ii;
+
+// 1D array -> 3D array
+   typedef real (*vla)[ArraySizeZ][ArraySizeY][ArraySizeX];
+   vla Array3D = ( vla )Array;
+
+
+// set the boundary values
+   for (int v=0; v<NVar_Flu; v++)
+   {
+      TFluVarIdx = TFluVarIdxList[v];
+
+      if ( TFluVarIdx == MOMX )
+      {
+         for (int k=Idx_Start[2]; k<=Idx_End[2]; k++)    {
+         for (int j=Idx_Start[1]; j<=Idx_End[1]; j++)    {
+         for (int i=Idx_Start[0]; i<=Idx_End[0]; i++)    {  ii = i_ref - i;
```

Comment:
The definitions of `j/k_ref` and `ii/jj/kk` are incorrect. In short, they should follow the same definitions in `Hydro_BoundaryCondition_Outflow.cpp` instead of `Hydro_BoundaryCondition_Reflecting.cpp`. I've put a correct version at [here](https://gist.github.com/hyschive/96273166d984851778bc8ff6976f44cf).